### ISSUE TITLE
Ability to specify WHERE clause on Prisma vector store

### ISSUE THREAD

**Issue by itedwards** [langchain-ai/langchainjs#692]
The current implementation allows for a SELECT clause to be provided on creation but not a WHERE clause. It would be great to apply filters on returned documents in a way that leverages generated Prisma types.

**Comment by kwkr** [comment]
Hey @itedwards! I wanted to give the implementation of it a shot. Could you provide me some more details for an example use case? As I'm looking at the current implementation, the SELECT clause is being constructed upon the creation by taking into account the columns passed. How would you like to apply the WHERE clause exactly? One way I can think of is passing some additional conditions to the `similaritySearch` methods. They could be used to construct the WHERE part which would affect the returned results in some way. Do I understand it correctly?

**Comment by stevenho0811** [comment]
I have created a new version of the "similaritySearchVectorWithScore" function for myself, and I added a "userId" parameter at the end of the args. The use case for this is that I don't want the entire message to be treated as the search content. Instead, I want to perform the search on the content spoken by each user based on their user id. Please refer to this for more information.

```ts
  async similaritySearchVectorWithScore<
    Model extends Message = Message,
    Column extends ModelColumns<Model> = ModelColumns<Model>,
  >(
    query: number[],
    k: number,
    userId: string,
  ): Promise<[Document<SimilarityModel<Message, Column>>, number][]> {
    const vectorQuery = `[${query.join(',')}]`;
    const articles = await this.prisma.$queryRaw<
      Array<SimilarityModel<Message, Column>>
    >`
      SELECT ${this.vector.selectSql}, ${this.vector.vectorColumnSql} <=> ${vectorQuery}::vector as "_distance" 
      FROM ${this.vector.tableSql}
      WHERE "userId" = ${userId}
      ORDER BY "_distance" ASC
      LIMIT ${k};
    `;

    const results: [Document<SimilarityModel<Message, Column>>, number][] = [];
    for (const article of articles) {
      if (
        article._distance != null &&
        article[this.vector.contentColumn] != null
      ) {
        results.push([
          new Document({
            pageContent: article[this.vector.contentColumn] as string,
            metadata: article,
          }),
          article._distance,
        ]);
      }
    }

    return results;
  }
```

**Comment by arsaizdihar** [comment]
I need this also. I think the implementation could be similar as the Supabase vector store metadata filtering. It is mentioned in the documents https://js.langchain.com/docs/modules/indexes/vector_stores/integrations/supabase#metadata-filtering

```ts
const result = await vectorStore.similaritySearch("Hello world", 1, {
    user_id: 3,
  });
```

**Comment by erictt** [comment]
I just build a PR for this: https://github.com/hwchase17/langchainjs/pull/1234

**Comment by devidw** [comment]
Thanks making filtering possible @erictt 

Wondering if there is a way to do the following with the current filter implementation?

```ts
const excludedIds = [1, 2, 3];  // The IDs you want to exclude

const result = await prisma.model.findMany({
  where: {
    id: {
      notIn: excludedIds,
    },
  },
});
```

Because there seems to be no `notIn` operator and the filter is an object and not an array so its not possible to pass as `[{id: {not: X}}, {id: {not: Y}}]`.

**Comment by devidw** [comment]
@disbelief you can not compare `uuid` and `text` types in postgres directly, would need to typecast the uuid to text for the comparison to work:

```sql
user_id::text = 'a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11'
```

But since langchain abstracts the raw sql call away, I can't currently think of a way to get this working with the built-in filters.

**Comment by disbelief** [comment]
@devidw yes I tried to "force" this type casting into the filter args in various ways without any luck.

Only way I could get an uuid-filtered vector similarity query to work was via `rawQueryUnsafe`:

```typescript
const docs = await prismaClient.$rawQueryUnsafe`
    SELECT
      "id",
      "text",
      "role",
      "user_id",
      "embedding" <=> '${vectorQuery}'::vector as "_distance"
    FROM "messages"
    WHERE "user_id" = '${user.id}'
    ORDER BY "_distance" ASC
    LIMIT 4;
  `;
```

Oddly no type casting was necessary for the `${user.id}`.

Sadly this means I'm unable to use the `PrismaVectorStore`.


Another gotcha with `PrismaVectorStore` is that Prisma's table name and column name mappings are not respected, but the typescript type signature does not allow you to specify the correct underlying table name if it differs from the model name in the schema.

My Postgres DB uses the snake_case_plural naming convention (eg. "messages" vs "Message", "user_id" vs "userId") so I had to use `// @ts-ignore` to pass the correct `tableName` and `filter` options.
